File: backend/routers/rundown.py
"""
Rundown Profiles API Router
API endpoints for contingency rundown profile generation and retrieval
"""

from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from pydantic import BaseModel, Field
from typing import Optional, List, Literal
from datetime import datetime
import logging

from supabase import Client
from auth.dependencies import get_current_user
from config.database import supabase
from services.rundown_generator import RundownGenerator, RundownGeneratorError

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/api/rundown",
    tags=["rundown"],
    responses={404: {"description": "Not found"}}
)


# ============================================
# Request/Response Models
# ============================================

class GenerateRequest(BaseModel):
    """Request body for profile generation"""
    project_id: Optional[str] = Field(
        None,
        description="Optional project ID. If not provided, generates for all projects."
    )
    profile_types: Optional[List[Literal["standard", "optimistic", "pessimistic"]]] = Field(
        default=["standard"],
        description="Types of profiles to generate"
    )
    scenario_name: Optional[str] = Field(
        default="baseline",
        description="Name of the scenario"
    )
    include_predictions: Optional[bool] = Field(
        default=True,
        description="Whether to include AI predictions"
    )


class GenerationError(BaseModel):
    """Error details from generation"""
    project_id: str
    project_name: Optional[str] = None
    error_type: str
    error_message: str
    timestamp: str


class GenerateResponse(BaseModel):
    """Response from profile generation"""
    execution_id: str
    status: Literal["started", "completed", "failed", "partial"]
    projects_processed: int
    profiles_created: int
    errors_count: int
    execution_time_ms: int
    errors: List[GenerationError]
    timestamp: str


class RundownProfile(BaseModel):
    """A single rundown profile data point"""
    id: str
    project_id: str
    month: str
    planned_value: float
    actual_value: float
    predicted_value: Optional[float] = None
    profile_type: str
    scenario_name: str
    created_at: datetime
    updated_at: datetime


class ProfilesResponse(BaseModel):
    """Response containing profile data"""
    project_id: str
    profiles: List[RundownProfile]
    total_count: int


class GenerationStatusResponse(BaseModel):
    """Status of profile generation"""
    last_execution_id: Optional[str] = None
    last_status: Optional[str] = None
    last_timestamp: Optional[datetime] = None
    projects_with_profiles: int
    total_profiles: int
    is_generating: bool = False


class RundownSummary(BaseModel):
    """Summary of rundown profiles for a project"""
    project_id: str
    project_name: str
    total_points: int
    start_month: str
    end_month: str
    current_planned: float
    current_actual: float
    current_predicted: Optional[float] = None
    variance: float
    variance_percentage: float
    is_over_budget: bool
    last_updated: datetime


# ============================================
# Background task for generation
# ============================================

async def run_generation_task(
    generator: RundownGenerator,
    project_id: Optional[str],
    profile_types: List[str],
    scenario_name: str,
    include_predictions: bool
):
    """Background task to run profile generation."""
    try:
        await generator.generate(
            project_id=project_id,
            profile_types=profile_types,
            scenario_name=scenario_name,
            include_predictions=include_predictions
        )
    except Exception as e:
        logger.error(f"Background generation task failed: {e}")


# ============================================
# API Endpoints
# ============================================

@router.post(
    "/generate",
    response_model=GenerateResponse,
    summary="Generate rundown profiles",
    description="""
    Generate or regenerate rundown profiles for projects.
    
    If project_id is provided, generates profiles for that specific project only.
    Otherwise, generates profiles for all active projects.
    
    The generation calculates:
    - Planned profile: Linear budget distribution over project duration
    - Actual profile: Adjusted values based on commitments and actuals
    - Predicted profile: AI-based predictions using linear regression
    """
)
async def generate_profiles(
    request: GenerateRequest,
    background_tasks: BackgroundTasks = None,
    current_user: dict = Depends(get_current_user)
):
    """Generate rundown profiles for one or all projects."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    generator = RundownGenerator(supabase)
    
    try:
        result = await generator.generate(
            project_id=request.project_id,
            profile_types=request.profile_types or ["standard"],
            scenario_name=request.scenario_name or "baseline",
            include_predictions=request.include_predictions if request.include_predictions is not None else True
        )
        
        return GenerateResponse(
            execution_id=result["execution_id"],
            status=result["status"],
            projects_processed=result["projects_processed"],
            profiles_created=result["profiles_created"],
            errors_count=result["errors_count"],
            execution_time_ms=result["execution_time_ms"],
            errors=[GenerationError(**e) for e in result.get("errors", [])],
            timestamp=result["timestamp"]
        )
        
    except RundownGeneratorError as e:
        logger.error(f"Generation error: {e}")
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        logger.error(f"Unexpected error in generation: {e}")
        raise HTTPException(status_code=500, detail="Profile generation failed")


@router.post(
    "/generate/async",
    summary="Start async profile generation",
    description="Start profile generation in the background and return immediately."
)
async def generate_profiles_async(
    request: GenerateRequest,
    background_tasks: BackgroundTasks,
    current_user: dict = Depends(get_current_user)
):
    """Start profile generation as a background task."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    generator = RundownGenerator(supabase)
    
    background_tasks.add_task(
        run_generation_task,
        generator,
        request.project_id,
        request.profile_types or ["standard"],
        request.scenario_name or "baseline",
        request.include_predictions if request.include_predictions is not None else True
    )
    
    return {
        "message": "Profile generation started",
        "execution_id": generator.execution_id
    }


@router.get(
    "/profiles/{project_id}",
    response_model=ProfilesResponse,
    summary="Get rundown profiles for a project",
    description="Retrieve all rundown profiles for a specific project."
)
async def get_project_profiles(
    project_id: str,
    profile_type: Optional[str] = Query(
        None,
        description="Filter by profile type"
    ),
    scenario_name: Optional[str] = Query(
        "baseline",
        description="Filter by scenario name"
    ),
    from_month: Optional[str] = Query(
        None,
        description="Start month (YYYYMM format)"
    ),
    to_month: Optional[str] = Query(
        None,
        description="End month (YYYYMM format)"
    ),
    current_user: dict = Depends(get_current_user)
):
    """Get rundown profiles for a specific project."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        query = supabase.table("rundown_profiles").select("*").eq(
            "project_id", project_id
        )
        
        if profile_type:
            query = query.eq("profile_type", profile_type)
        if scenario_name:
            query = query.eq("scenario_name", scenario_name)
        if from_month:
            query = query.gte("month", from_month)
        if to_month:
            query = query.lte("month", to_month)
            
        query = query.order("month")
        
        response = query.execute()
        profiles = response.data or []
        
        return ProfilesResponse(
            project_id=project_id,
            profiles=[RundownProfile(**p) for p in profiles],
            total_count=len(profiles)
        )
        
    except Exception as e:
        logger.error(f"Error fetching profiles for project {project_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch profiles")


@router.get(
    "/profiles",
    summary="Get rundown profiles for multiple projects",
    description="Retrieve rundown profiles for multiple projects at once."
)
async def get_multiple_profiles(
    project_ids: List[str] = Query(
        ...,
        description="List of project IDs"
    ),
    profile_type: Optional[str] = Query(
        "standard",
        description="Filter by profile type"
    ),
    scenario_name: Optional[str] = Query(
        "baseline",
        description="Filter by scenario name"
    ),
    current_user: dict = Depends(get_current_user)
):
    """Get rundown profiles for multiple projects."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        query = supabase.table("rundown_profiles").select("*").in_(
            "project_id", project_ids
        )
        
        if profile_type:
            query = query.eq("profile_type", profile_type)
        if scenario_name:
            query = query.eq("scenario_name", scenario_name)
            
        query = query.order("project_id").order("month")
        
        response = query.execute()
        profiles = response.data or []
        
        # Group by project
        grouped = {}
        for profile in profiles:
            pid = profile["project_id"]
            if pid not in grouped:
                grouped[pid] = []
            grouped[pid].append(profile)
            
        return {
            "profiles_by_project": grouped,
            "total_profiles": len(profiles),
            "projects_count": len(grouped)
        }
        
    except Exception as e:
        logger.error(f"Error fetching multiple profiles: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch profiles")


@router.get(
    "/summary/{project_id}",
    response_model=RundownSummary,
    summary="Get rundown summary for a project",
    description="Get a summary of the rundown profiles for a project."
)
async def get_project_summary(
    project_id: str,
    scenario_name: Optional[str] = Query("baseline"),
    current_user: dict = Depends(get_current_user)
):
    """Get rundown summary for a specific project."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        # Fetch project info
        project_response = supabase.table("projects").select(
            "id, name"
        ).eq("id", project_id).single().execute()
        
        project = project_response.data
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
            
        # Fetch profiles
        profiles_response = supabase.table("rundown_profiles").select("*").eq(
            "project_id", project_id
        ).eq("scenario_name", scenario_name).order("month").execute()
        
        profiles = profiles_response.data or []
        
        if not profiles:
            raise HTTPException(
                status_code=404, 
                detail="No rundown profiles found for this project"
            )
            
        # Calculate summary
        current_month = datetime.utcnow().strftime("%Y%m")
        current_profile = next(
            (p for p in profiles if p["month"] == current_month),
            profiles[-1]  # Use latest if current month not found
        )
        
        current_planned = current_profile["planned_value"]
        current_actual = current_profile["actual_value"]
        variance = current_actual - current_planned
        variance_pct = (variance / current_planned * 100) if current_planned else 0
        
        return RundownSummary(
            project_id=project_id,
            project_name=project["name"],
            total_points=len(profiles),
            start_month=profiles[0]["month"],
            end_month=profiles[-1]["month"],
            current_planned=current_planned,
            current_actual=current_actual,
            current_predicted=current_profile.get("predicted_value"),
            variance=variance,
            variance_percentage=round(variance_pct, 2),
            is_over_budget=variance > 0,
            last_updated=profiles[-1]["updated_at"]
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting summary for project {project_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to get summary")


@router.get(
    "/status",
    response_model=GenerationStatusResponse,
    summary="Get generation status",
    description="Get the current status of profile generation."
)
async def get_generation_status(
    current_user: dict = Depends(get_current_user)
):
    """Get the current generation status."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        # Get last execution
        last_log = supabase.table("rundown_generation_logs").select("*").order(
            "created_at", desc=True
        ).limit(1).execute()
        
        last_entry = last_log.data[0] if last_log.data else None
        
        # Count projects with profiles
        profiles_count = supabase.table("rundown_profiles").select(
            "project_id", count="exact"
        ).execute()
        
        distinct_projects = supabase.rpc(
            "count_distinct_projects_with_profiles"
        ).execute() if False else {"count": 0}  # Use raw count if RPC not available
        
        # Count total profiles
        total = supabase.table("rundown_profiles").select("*", count="exact").execute()
        
        return GenerationStatusResponse(
            last_execution_id=last_entry["execution_id"] if last_entry else None,
            last_status=last_entry["status"] if last_entry else None,
            last_timestamp=last_entry["created_at"] if last_entry else None,
            projects_with_profiles=len(set(p["project_id"] for p in (profiles_count.data or []))),
            total_profiles=total.count or 0,
            is_generating=False
        )
        
    except Exception as e:
        logger.error(f"Error getting generation status: {e}")
        raise HTTPException(status_code=500, detail="Failed to get status")


@router.get(
    "/logs",
    summary="Get generation logs",
    description="Get recent profile generation logs."
)
async def get_generation_logs(
    limit: int = Query(20, ge=1, le=100),
    execution_id: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get generation logs."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        query = supabase.table("rundown_generation_logs").select("*")
        
        if execution_id:
            query = query.eq("execution_id", execution_id)
            
        query = query.order("created_at", desc=True).limit(limit)
        
        response = query.execute()
        
        return {
            "logs": response.data or [],
            "count": len(response.data or [])
        }
        
    except Exception as e:
        logger.error(f"Error fetching generation logs: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch logs")


@router.delete(
    "/profiles/{project_id}",
    summary="Delete rundown profiles",
    description="Delete all rundown profiles for a project."
)
async def delete_project_profiles(
    project_id: str,
    scenario_name: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Delete rundown profiles for a project."""
    if supabase is None:
        raise HTTPException(status_code=503, detail="Database service unavailable")

    try:
        query = supabase.table("rundown_profiles").delete().eq(
            "project_id", project_id
        )
        
        if scenario_name:
            query = query.eq("scenario_name", scenario_name)
            
        response = query.execute()
        
        return {
            "message": "Profiles deleted successfully",
            "project_id": project_id,
            "scenario_name": scenario_name
        }
        
    except Exception as e:
        logger.error(f"Error deleting profiles for project {project_id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to delete profiles")

File: types/rundown.ts
// Rundown Profile Type Definitions
// Contingency Rundown Profiles for Costbook

/**
 * Profile type variants
 */
export type ProfileType = 'standard' | 'optimistic' | 'pessimistic'

/**
 * Generation status
 */
export type GenerationStatus = 'started' | 'completed' | 'failed' | 'partial'

/**
 * Scenario adjustment type
 */
export type AdjustmentType = 'percentage' | 'absolute'

/**
 * A single rundown profile data point
 */
export interface RundownProfile {
  id: string
  project_id: string
  /** Month in YYYYMM format */
  month: string
  /** Planned cumulative budget value */
  planned_value: number
  /** Actual cumulative spend value */
  actual_value: number
  /** AI-predicted cumulative value (null for past months) */
  predicted_value: number | null
  /** Profile type variant */
  profile_type: ProfileType
  /** Named scenario for what-if analysis */
  scenario_name: string
  created_at: string
  updated_at: string
}

/**
 * Summary of rundown profiles for a project
 */
export interface RundownProfileSummary {
  project_id: string
  project_name: string
  /** Total number of profile data points */
  total_points: number
  /** First month in YYYYMM format */
  start_month: string
  /** Last month in YYYYMM format */
  end_month: string
  /** Current month's planned value */
  current_planned: number
  /** Current month's actual value */
  current_actual: number
  /** Current month's predicted value */
  current_predicted: number | null
  /** Variance (actual - planned) */
  variance: number
  /** Variance percentage */
  variance_percentage: number
  /** Whether project is over budget */
  is_over_budget: boolean
  /** Last profile update timestamp */
  last_updated: string
}

/**
 * Rundown scenario for what-if analysis
 */
export interface RundownScenario {
  id: string
  project_id: string
  name: string
  description?: string
  /** Type of adjustment */
  adjustment_type: AdjustmentType
  /** Adjustment value (percentage or absolute amount) */
  adjustment_value: number
  is_active: boolean
  created_by?: string
  created_at: string
  updated_at: string
}

/**
 * Generation error details
 */
export interface GenerationError {
  project_id: string
  project_name?: string
  error_type: string
  error_message: string
  timestamp: string
}

/**
 * Result from profile generation
 */
export interface GenerationResult {
  execution_id: string
  status: GenerationStatus
  /** Total projects processed */
  projects_processed: number
  /** Total profiles created/updated */
  profiles_created: number
  /** Number of errors */
  errors_count: number
  /** Execution time in milliseconds */
  execution_time_ms: number
  /** List of errors if any */
  errors: GenerationError[]
  /** Timestamp of execution */
  timestamp: string
}

/**
 * Generation log entry
 */
export interface GenerationLog {
  id: string
  execution_id: string
  project_id?: string
  status: GenerationStatus
  message?: string
  projects_processed: number
  profiles_created: number
  errors_count: number
  execution_time_ms?: number
  created_at: string
}

/**
 * Input for generating rundown profiles
 */
export interface GenerateProfilesInput {
  /** Optional specific project ID (if not provided, generates for all) */
  project_id?: string
  /** Profile types to generate */
  profile_types?: ProfileType[]
  /** Whether to recalculate predictions */
  include_predictions?: boolean
  /** Specific scenario to generate */
  scenario_name?: string
}

/**
 * Query options for fetching profiles
 */
export interface RundownProfilesQuery {
  project_id: string
  profile_type?: ProfileType
  scenario_name?: string
  /** Start month (YYYYMM) */
  from_month?: string
  /** End month (YYYYMM) */
  to_month?: string
}

/**
 * Chart data point for visualization
 */
export interface RundownChartPoint {
  /** Month in YYYYMM format */
  month: string
  /** Month label for display (e.g., "Jan 24") */
  label: string
  /** Planned value */
  planned: number
  /** Actual value */
  actual: number
  /** Predicted value (for future months) */
  predicted: number | null
  /** Whether this is the current month */
  isCurrent: boolean
  /** Whether this is a future month */
  isFuture: boolean
}

/**
 * Prediction analysis result
 */
export interface PredictionAnalysis {
  /** Trend direction */
  trend: 'increasing' | 'stable' | 'decreasing'
  /** Prediction confidence (0-100) */
  confidence: number
  /** Predicted final value at project end */
  predicted_final: number
  /** Variance from planned at project end */
  predicted_variance: number
  /** Whether prediction exceeds planned by threshold */
  exceeds_threshold: boolean
  /** Percentage over planned (if exceeds) */
  threshold_percentage: number
}

/**
 * Rundown profile status for UI
 */
export interface RundownStatus {
  /** Last generation timestamp */
  last_generated?: string
  /** Next scheduled generation */
  next_scheduled?: string
  /** Whether profiles are stale (>24h old) */
  is_stale: boolean
  /** Whether generation is in progress */
  is_generating: boolean
  /** Number of projects with profiles */
  projects_with_profiles: number
  /** Total profile count */
  total_profiles: number
  /** Any active errors */
  has_errors: boolean
  /** Error count */
  error_count: number
}

/**
 * Configuration for rundown generation
 */
export interface RundownConfig {
  /** Cron schedule (default: daily at 02:00 UTC) */
  cron_schedule: string
  /** Number of months to predict ahead */
  prediction_months: number
  /** Threshold for prediction warnings (percentage) */
  warning_threshold: number
  /** Enable real-time updates */
  realtime_enabled: boolean
  /** Debounce time for real-time updates (ms) */
  realtime_debounce_ms: number
}

/**
 * Default configuration values
 */
export const DEFAULT_RUNDOWN_CONFIG: RundownConfig = {
  cron_schedule: '0 2 * * *', // Daily at 02:00 UTC
  prediction_months: 6,
  warning_threshold: 10, // 10% over planned
  realtime_enabled: true,
  realtime_debounce_ms: 2000
}

/**
 * Helper to format month string to display label
 */
export function formatMonthLabel(month: string): string {
  const year = month.substring(2, 4)
  const monthNum = parseInt(month.substring(4, 6), 10)
  const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
                      'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
  return `${monthNames[monthNum - 1]} ${year}`
}

/**
 * Get current month in YYYYMM format
 */
export function getCurrentMonth(): string {
  const now = new Date()
  const year = now.getFullYear()
  const month = (now.getMonth() + 1).toString().padStart(2, '0')
  return `${year}${month}`
}

/**
 * Compare two YYYYMM month strings
 * Returns negative if a < b, positive if a > b, 0 if equal
 */
export function compareMonths(a: string, b: string): number {
  return parseInt(a, 10) - parseInt(b, 10)
}

/**
 * Check if a month is in the future
 */
export function isFutureMonth(month: string): boolean {
  return compareMonths(month, getCurrentMonth()) > 0
}

/**
 * Check if a month is the current month
 */
export function isCurrentMonth(month: string): boolean {
  return month === getCurrentMonth()
}

/**
 * Calculate variance between actual and planned
 */
export function calculateVariance(actual: number, planned: number): {
  absolute: number
  percentage: number
  isOver: boolean
} {
  const absolute = actual - planned
  const percentage = planned !== 0 ? (absolute / planned) * 100 : 0
  return {
    absolute,
    percentage,
    isOver: absolute > 0
  }
}

/**
 * Convert profiles to chart data points
 */
export function profilesToChartData(profiles: RundownProfile[]): RundownChartPoint[] {
  const currentMonth = getCurrentMonth()
  
  return profiles
    .sort((a, b) => compareMonths(a.month, b.month))
    .map(profile => ({
      month: profile.month,
      label: formatMonthLabel(profile.month),
      planned: profile.planned_value,
      actual: profile.actual_value,
      predicted: profile.predicted_value,
      isCurrent: profile.month === currentMonth,
      isFuture: compareMonths(profile.month, currentMonth) > 0
    }))
}
